### DISCUSSION TITLE
Using registryAliases with datasource "docker" leads to "depName mismatch" errors

### DISCUSSION THREAD

**Discussion by ChristianCiach** [renovatebot/renovate#23188] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.0.3

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

Preface: I never tried this with Renovate v35 before, but according to previous issues/discussions, this should've worked in the past.

I am trying to get the `registryAliases` for docker images to work. 

**I've prepared a minimal reproduction**: https://github.com/ChristianCiach/renovate-v36-depNameMismatch  -- Please also see the executed pipeline for debug logs.

Long story short: I have this Dockerfile:

```dockerfile
FROM docker.io/library/alpine:3.17
```

Because of rate-limiting, I want to rewrite `docker.io` to our internal docker-hub mirror. For the sake of reproduction, let's use the public mirror at AWS:

```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": ["config:base"],
    "packageRules": [
        {
            "matchDatasources": ["docker"],
            "registryAliases": {
                "docker.io": "public.ecr.aws/docker"
            }
        }
    ]
}
```

After reading a lot of issues, discussion and the source code, I am reasonable sure that this is expected to work. But unfortunately Renovate refuses to create a branch:

```
DEBUG: Found match at index 5 (repository=ChristianCiach/renovate-v36-depNameMismatch, packageFile=Dockerfile, branch=renovate/docker.io-library-alpine-3.x)
       "depName": "docker.io/library/alpine"
DEBUG: depName mismatch (repository=ChristianCiach/renovate-v36-depNameMismatch, packageFile=Dockerfile, branch=renovate/docker.io-library-alpine-3.x)
       "manager": "dockerfile",
       "currentDepName": "docker.io/library/alpine",
       "newDepName": "public.ecr.aws/docker/library/alpine"
 WARN: Error updating branch: update failure (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
```

Sad face.

### Relevant debug logs

<details><summary>Logs</summary>

```
 DEBUG: syncBranchState() (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: syncBranchState(): Branch cache not found, creating minimal branchState (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: getBranchPr(renovate/docker.io-library-alpine-3.x) (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: findPr(renovate/docker.io-library-alpine-3.x, undefined, open) (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: findPr(renovate/docker.io-library-alpine-3.x, undefined, closed) (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: branchExists=false (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: dependencyDashboardCheck=undefined (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: Check for closed PR because recreating closed PRs is disabled. (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: findPr(renovate/docker.io-library-alpine-3.x, Update docker.io/library/alpine Docker tag to v3.18, !open) (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: prAlreadyExisted=false (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: Checking schedule(at any time, null) (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: No schedule defined (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: Branch needs creating (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: Using reuseExistingBranch: false (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: Setting current branch to main (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: latest commit (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
       "branchName": "main",
       "latestCommitDate": "2023-07-05T21:30:30+02:00"
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=false (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: Starting search at index 5 (repository=ChristianCiach/renovate-v36-depNameMismatch, packageFile=Dockerfile, branch=renovate/docker.io-library-alpine-3.x)
       "depName": "docker.io/library/alpine"
DEBUG: Found match at index 5 (repository=ChristianCiach/renovate-v36-depNameMismatch, packageFile=Dockerfile, branch=renovate/docker.io-library-alpine-3.x)
       "depName": "docker.io/library/alpine"
DEBUG: depName mismatch (repository=ChristianCiach/renovate-v36-depNameMismatch, packageFile=Dockerfile, branch=renovate/docker.io-library-alpine-3.x)
       "manager": "dockerfile",
       "currentDepName": "docker.io/library/alpine",
       "newDepName": "public.ecr.aws/docker/library/alpine"
 WARN: Error updating branch: update failure (repository=ChristianCiach/renovate-v36-depNameMismatch, branch=renovate/docker.io-library-alpine-3.x)
DEBUG: getPrCache() (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: getPrCache() (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: ensureDependencyDashboard() (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: Ensuring Dependency Dashboard (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: repository problems (repository=ChristianCiach/renovate-v36-depNameMismatch)
       "repoProblems": ["WARN: Error updating branch: update failure"]
DEBUG: ensureIssue(Dependency Dashboard) (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: Patching issue (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: Issue updated (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: Removing any stale branches (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: config.repoIsOnboarded=true (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: No renovate branches found (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: PackageFiles.clear() - Package files deleted (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: Branch summary (repository=ChristianCiach/renovate-v36-depNameMismatch)
       "cacheModified": undefined,
       "baseBranches": [{"branchName": "main", "sha": "3b1fb33a1788a015f960a141456aa701528be0b7"}],
       "branches": [],
       "defaultBranch": "main",
       "inactiveBranches": [
         "renovate/renovatebot-github-action-38.x",
         "renovate/docker.io-library-alpine-3.x"
       ]
DEBUG: branches info extended (repository=ChristianCiach/renovate-v36-depNameMismatch)
       "branchesInformation": [
         {
           "branchName": "renovate/renovatebot-github-action-38.x",
           "prNo": null,
           "prTitle": "Update renovatebot/github-action action to v38.1.9",
           "result": "no-work",
           "upgrades": [
             {
               "datasource": "github-tags",
               "depName": "renovatebot/github-action",
               "displayPending": "",
               "fixedVersion": "v38.1.6",
               "currentVersion": "v38.1.6",
               "currentValue": "v38.1.6",
               "newValue": "v38.1.9",
               "newVersion": "v38.1.9",
               "packageFile": ".github/workflows/main.yml",
               "updateType": "patch",
               "packageName": "renovatebot/github-action"
             }
           ]
         },
         {
           "branchName": "renovate/docker.io-library-alpine-3.x",
           "prNo": null,
           "prTitle": "Update docker.io/library/alpine Docker tag to v3.18",
           "result": "error",
           "upgrades": [
             {
               "datasource": "docker",
               "depName": "docker.io/library/alpine",
               "displayPending": "",
               "fixedVersion": "3.17",
               "currentVersion": "3.17",
               "currentValue": "3.17",
               "newValue": "3.18",
               "newVersion": "3.18",
               "packageFile": "Dockerfile",
               "updateType": "minor",
               "packageName": "docker.io/library/alpine"
             }
           ]
         }
       ]
DEBUG: Renovate repository PR statistics (repository=ChristianCiach/renovate-v36-depNameMismatch)
       "stats": {"total": 0, "open": 0, "closed": 0, "merged": 0}
DEBUG: Repository result: done, status: onboarded, enabled: true, onboarded: true (repository=ChristianCiach/renovate-v36-depNameMismatch)
DEBUG: Repository timing splits (milliseconds) (repository=ChristianCiach/renovate-v36-depNameMismatch)
       "splits": {"init": 1656, "extract": 517, "lookup": 9764, "onboarding": 3, "update": 931},
       "total": 13707
DEBUG: Package cache statistics (repository=ChristianCiach/renovate-v36-depNameMismatch)
       "get": {"count": 6, "avgMs": 3, "medianMs": 0, "maxMs": 16},
       "set": {"count": 2, "avgMs": 7, "medianMs": 3, "maxMs": 10}
DEBUG: http statistics (repository=ChristianCiach/renovate-v36-depNameMismatch)
       "urls": {
         "https://api.github.com/graphql (POST,200)": 32,
         "https://api.github.com/repos/ChristianCiach/renovate-v36-depNameMismatch/issues/1 (GET,200)": 2,
         "https://api.github.com/repos/ChristianCiach/renovate-v36-depNameMismatch/issues/1 (PATCH,200)": 1,
         "https://api.github.com/repos/ChristianCiach/renovate-v36-depNameMismatch/pulls (GET,200)": 1,
         "https://auth.docker.io/token (GET,200)": 1,
         "https://index.docker.io/v2/ (GET,401)": 1,
         "https://index.docker.io/v2/library/alpine/blobs/sha256:c1aabb73d2339c5ebaa3681de2e9d9c18d57485045a4e311d9f8004bec208d67 (GET,200)": 1,
         "https://index.docker.io/v2/library/alpine/manifests/latest (GET,200)": 1,
         "https://index.docker.io/v2/library/alpine/manifests/sha256:25fad2a32ad1f6f510e528448ae1ec69a28ef81916a004d3629874104f8a7f70 (GET,200)": 1,
         "https://index.docker.io/v2/library/alpine/tags/list (GET,200)": 1,
         "https://index.docker.io/v2/library/alpine/tags/list (GET,401)": 1
       },
       "hostStats": {
         "api.github.com": {"requestCount": 36, "requestAvgMs": 306, "queueAvgMs": 0},
         "auth.docker.io": {"requestCount": 1, "requestAvgMs": 34, "queueAvgMs": 0},
         "index.docker.io": {"requestCount": 6, "requestAvgMs": 62, "queueAvgMs": 0}
       },
       "totalRequests": 43
DEBUG: Package lookup durations (repository=ChristianCiach/renovate-v36-depNameMismatch)
       "docker": {"count": 1, "averageMs": 539, "totalMs": 539, "maximumMs": 539},
       "github-tags": {"count": 2, "averageMs": 5139, "totalMs": 10278, "maximumMs": 9669}
DEBUG: dns cache (repository=ChristianCiach/renovate-v36-depNameMismatch)
       "hosts": []
 INFO: Repository finished (repository=ChristianCiach/renovate-v36-depNameMismatch)
       "cloned": true,
       "durationMs": 13707
DEBUG: Checking file package cache for expired items
DEBUG: Deleted 0 of 2 file cached entries in 6ms
DEBUG: Renovate exiting
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
Can you use `FROM alpine:3.17` instead? I think in that case you can use `registryUrls`

**Reply by ChristianCiach** [reply]
Well, yes, `registryUrls` works and I've configured that, too. 

But unfortunately we have **many** use-cases where we absolute need to use the fully qualified image name (including the registry) when referencing an image. I could write a wall of text for why we need that, but let me give you the gist of it:

* Many of our systems are air-gapped and we need our image-references to be "relocatable" by just adding a prefix to the image name. For example, if our docker-hub mirror was `my.mirror.local/docker.io` I need to be able to automatically prefix the image reference from the dockerfile with the mirror hostname: `docker.io/library/alpine:3.17` would become `my.mirror.local/docker.io/library/alpine:3.17`, but `alpine:3.17` would become `my.mirror.local/alpine:3.17`, which is wrong.
  * Actually, many of our Dockerfiles are templated like `FROM {{ mirror_prefix }}/library/alpine`. Since the `mirror_prefix` can contain a prefix-path like `/docker`, we need to fully qualify the image reference. This is pretty much the same as with the `public.ecr.aws/docker` mirror mentioned in the description.
* The assumption that the registry-less image reference like `alpine:3.17` *necessarily* refers to docker-hub is not universally true. This is true for moby-based container runtimes (docker, containerd), but this is actually configurable for container runtimes by other vendors (podman, OpenShift (CRI-O)). We *do* have environments where the default-registry has been configured to *not* be docker-hub. In these environments we **need** to explicitly reference `docker.io/` for images that should be pulled from docker hub.

TL;DR: `registryAliases` and `registryUrls` are used for difference use-cases. We want to use both, but in different situations.

**Reply by rarkins** [reply]
Can you set up a public reproduction? You could use gel public registries, Eg Docker.io and quay.io

**Reply by ChristianCiach** [reply]
I've linked a public reproduction directly in the description when I created this discussion :)

**Comment by rarkins** [comment]
When I run against the reproduction repo, I can see that Renovate is accessing docker.io directly:

```
         "https://auth.docker.io/token (GET,200)": 1,
         "https://index.docker.io/v2/ (GET,401)": 1,
         "https://index.docker.io/v2/library/alpine/blobs/sha256:c1aabb73d2339c5ebaa3681de2e9d9c18d57485045a4e311d9f8004bec208d67 (GET,200)": 1,
         "https://index.docker.io/v2/library/alpine/manifests/latest (GET,200)": 1,
         "https://index.docker.io/v2/library/alpine/manifests/sha256:25fad2a32ad1f6f510e528448ae1ec69a28ef81916a004d3629874104f8a7f70 (GET,200)": 1,
         "https://index.docker.io/v2/library/alpine/tags/list (GET,200)": 1,
         "https://index.docker.io/v2/library/alpine/tags/list (GET,401)": 1
```

**Reply by rarkins** [reply, marked as answer]
Check out https://github.com/renovate-reproductions/23188/pull/3 and the config change I made.

registryAliases are applied at time of dependency extraction, so making a package rule based on datasource=docker is too late for it to be applied.

**Reply by ChristianCiach** [reply]
Thanks, that seems to be it!

For clarification: `dockerfile {...}` should be equivalent to `"packageRules": [{"matchManagers": ["dockerfile"], ...}]`, right? I am asking because `dockerfile {...}` is not parts of the docs (anymore).

**Reply by ChristianCiach** [reply]
I am not a big fan of the fact that `registryAliases` actually changes the package-name that is used inside the PRs. I would have expected that Renovate transparently changes the name for the purpose of registry requests (hence registry**Aliases**).

But if this is the way that it is supposed to work, then I guess there is nothing for me to complain  about :)

**Reply by rarkins** [reply]
registryAliases is kind of a hack at this point, I agree with you but also not sure how much effort it will be to change
